tests/test_smoketest.py

```diff
@@ -12,9 +12,26 @@
 import cohortextractor
 
 
+@pytest.fixture
+def study(request):
+    study_name = request.function.__name__
+    study_path = os.path.join(os.path.dirname(__file__), "studies", study_name)
+    cohortextractor_path = os.path.dirname(os.path.dirname(cohortextractor.__file__))
+
+    def _study(*args):
+        subprocess.run(
+            [sys.executable, "-m", "cohortextractor.cohortextractor", *args],
+            check=True,
+            cwd=study_path,
+            env=dict(os.environ, PYTHONPATH=cohortextractor_path),
+        )
```

Comment:
I feel that returning a closure here is a slightly overcomplicated way of achieving this, especially combined with using `request.function.__name__`.

I'd be inclined to add a `fixture_name` argument to `_cohortextractor` like:
```py
def _cohortextractor(fxiture_name, *args):
    fixture_path = os.path.join(os.path.dirname(__file__), "fixtures", fixture_name)
   ...
```

This does require supplying the fixture name each time you call it in the tests, but that's not loads of extra typing and I think it would be a lot clearer what's going on.

Reply:
I started with that, but I didn't like passing the fixture/study name as a bare string at the start of the list of args because it's different to all the other args (they're all arguments to the cohortextrator).

```py
    _cohortextrator(
        "test_smoketest",
        "generate_cohort",
        "--expectations-population",
        "100",
        "--index-date-range",
        "2020-01-01 to 2020-04-01 by month",
        "--output-dir",
        tmp_path,
        "--output-format",
        format,
    )
```

And you can't do this, because it's not valid syntax (positional args after keyword arg):

```py
    _cohortextrator(
        study="test_smoketest",
        "generate_cohort",
        "--expectations-population",
        "100",
        "--index-date-range",
        "2020-01-01 to 2020-04-01 by month",
        "--output-dir",
        tmp_path,
        "--output-format",
        format,
    )
```

So I think it would have to be something like this:

```py
    _cohortextrator(
        study="test_smoketest",
        args=[
            "generate_cohort",
            "--expectations-population",
            "100",
            "--index-date-range",
            "2020-01-01 to 2020-04-01 by month",
            "--output-dir",
            tmp_path,
            "--output-format",
            format,
        ]
    )
```

I thought that was getting a bit messy and the fixture route seemed like a way out -- and allowed me to get rid of the test name/study name duplication.

Passing fixtures that are callables seems like a fairly usual pytest pattern. You're right that the combination with `request.function.__name__` makes it a bit non-obvious what the implementation is, but the clean surface seems worth it to me. I can be talked out of it, though.

Which option do you prefer?

Reply:
Ah yes, I had the same thought about mixing the arguments like that but to my mind that was more of a slight infelicity than a deal breaker. Overall though, I feel like this suggestion is the simplest and clearest to future readers:
```py
    _cohortextrator(
        study="test_smoketest",
        args=[
            ...
        ]
    )
```

To my mind, avoiding `request.function.__name__` is the main thing because I fear that someone will come along and change a test name to make it more accurate, or try to duplicate and tweak an existing test and then get very confused when it doesn't work. But I can't see an obvious way of supplying the name explicitly while using the callable fixture pattern.

You could keep the closure and have a function like:
```py
_cohortextractor = make_cohortextractor_for_study("fixture_name_here")
```
which you call explicitly at the top of each test function. But then that's a lot of machinery essentially just to avoid indenting `args`!

Reply:
You _could_ have it blow up with a very clear error message if the study
it's looking for doesn't exist.

Or `study("test_smoketest")(...cohortextractor args...)`, but I am going to
hazard a guess you don't like that either (and I don't much).

I'm happy to go with the one you suggest above, though.